src/java.base/macosx/native/libjava/java_props_macosx.c

```diff
@@ -222,37 +222,23 @@ char *setupMacOSXLocale(int cat) {
     }
 }
 
-// 10.9 SDK does not include the NSOperatingSystemVersion struct.
-// For now, create our own
-typedef struct {
-        NSInteger majorVersion;
-        NSInteger minorVersion;
-        NSInteger patchVersion;
-} OSVerStruct;
-
 void setOSNameAndVersion(java_props_t *sprops) {
     // Hardcode os_name, and fill in os_version
     sprops->os_name = strdup("Mac OS X");
 
     char* osVersionCStr = NULL;
-    // Mac OS 10.9 includes the [NSProcessInfo operatingSystemVersion] function,
-    // but it's not in the 10.9 SDK.  So, call it via objc_msgSend_stret.
-    if ([[NSProcessInfo processInfo] respondsToSelector:@selector(operatingSystemVersion)]) {
-        OSVerStruct (*procInfoFn)(id rec, SEL sel) = (OSVerStruct(*)(id, SEL))objc_msgSend_stret;
-        OSVerStruct osVer = procInfoFn([NSProcessInfo processInfo],
-                                       @selector(operatingSystemVersion));
-        NSString *nsVerStr;
-        if (osVer.patchVersion == 0) { // Omit trailing ".0"
-            nsVerStr = [NSString stringWithFormat:@"%ld.%ld",
-                    (long)osVer.majorVersion, (long)osVer.minorVersion];
-        } else {
-            nsVerStr = [NSString stringWithFormat:@"%ld.%ld.%ld",
-                    (long)osVer.majorVersion, (long)osVer.minorVersion, (long)osVer.patchVersion];
+    // Mac OS 10.2 includes the [NSProcessInfo operatingSystemVersionString] function,
+    // but it's not in the 10.2 SDK.  So, call it via objc_msgSend_stret.
+    if ([[NSProcessInfo processInfo] respondsToSelector:@selector(operatingSystemVersionString)]) {
+        NSString* (*procInfoFn)(id rec, SEL sel) = (NSString* (*)(id, SEL))objc_msgSend;
+        NSString *nsVerStr = procInfoFn([NSProcessInfo processInfo],
+                                       @selector(operatingSystemVersionString));
+        if (nsVerStr != NULL) {
+            // Copy out the char*
+            osVersionCStr = strdup([nsVerStr UTF8String]);
```

Comment:
This change looks good, i like it more than ours, but it discloses something you can find only with xcode12beta

Reply:
xcode12 is also available on macOS-x86 and all of this code can be cross-compiled.

Reply:
But xcode12 for macos doesn't have sdk for macos 11.0 (big sur), it only has sdk for 10.15. therefore it can't cross-compile for macarm.
I like this part, but the reason we have to do this can't be disclosed until xcode12.2 for universal apps is released.

Reply:
I don't know about xcode12.2, but xcode12 beta 6 _does_ have the SDK for big sur, including universal libs.

 

```console
$ xcodebuild -version
Xcode 12.0
Build version 12A8189n
$ jq '.SupportedTargets.macosx' < `xcode-select -p`/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/SDKSettings.json
{
  "MinimumDeploymentTarget": "10.9",
  "PlatformFamilyName": "macOS",
  "LLVMTargetTripleEnvironment": "",
  "LLVMTargetTripleSys": "macosx",
  "MaximumDeploymentTarget": "11.0.99",
  "LLVMTargetTripleVendor": "apple",
  "BuildVersionPlatformID": "1",
  "DeploymentTargetSettingName": "MACOSX_DEPLOYMENT_TARGET",
  "Archs": [
    "x86_64",
    "x86_64h",
    "arm64",
    "arm64e"
  ],
  "ValidDeploymentTargets": [
    "10.9",
    "10.10",
    "10.11",
    "10.12",
    "10.13",
    "10.14",
    "10.15",
    "11.0"
  ],
  "DefaultDeploymentTarget": "11.0"
}